SLM Hamamatsu

ichristen opened the pull request:
Add SDK support for Hamamatsu SLMs.

@ninovarde provided python scripts used to control LCOS-SLM X15213-02 by Hamamatsu Description of the files:
 - "L55-0491_USB_Control_SDK.pdf": file given from Hamamatsu containing all the documentation of the functions included in the DLL. 
 - "hamamatsu_slm_interpreter.py":   Python script that defines the class "Hamamatsu" used to controll the device. 
 - "hamamatsu_slm_class.py": Python script that implements the functions described in the PDF file. Consider this more as an example, as it was written based on our needs 
    and it is also affected by library updates.

ichristen reviewed (commented):
Hi @ninovarde , please see my comments.

As a general comment, any public functions will also need to have docstrings in the style of the rest of the package. In part because of this, my preference is to remove repeated or superfluous functions and focus on comprehensively documenting the functions needed to implement the SDK.

Let me know what you think!

ninovarde commented:
Dear Ian,Thank you for your comments. I will carefully read them and make the necessary adjustments to the scripts. I’ll update you in a couple of days, as I’m currently not able to spend much time at the computer.Thank you for your time.Best regards,Antonino VardèSent from my iPhoneOn 3 Apr 2025, at 21:05, Ian Christen ***@***.***> wrote:﻿
@ichristen commented on this pull request.

Hi @ninovarde , please see my comments.
As a general comment, any public functions will also need to have docstrings in the style of the rest of the package. In part because of this, my preference is to remove repeated or superfluous functions and focus on comprehensively documenting the functions needed to implement the SDK.
Let me know what you think!

In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            blaze_angle = None, # in degree
+            mask_radius = None # in pixel

If we add these arguments, it should be in the superclass SLM. This is already sortof implementable by setting the phase correction to a blaze. Would have to check, but setting part of the phase correction to np.nan might play the same role as the mask.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            frame_mem_dim = 819,
+            head_serial_numb = 'LSH0804453\x00',
+            bID_size = 1,

Let's try to keep the number of arguments down to the minimum. Otherwise, the package becomes unmaintainable. See other SLM classes for what minimum looks like.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.width = resolution[0]
+            self.height = resolution[1]

Removed in 0.1.0



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            temp =  ham.Check_Temp(self.ID)
+            self.head_temperature = temp[0]
+            self.controller_temperature = temp[1]
+            
+            self.led_status = ham.Check_LED(self.ID)

Let's not have static cached attributes here. Instead, the user should poll the hardware for fresh variables.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.blaze_angle = blaze_angle # angle at which the produced patterns are shifted from the zero order
+        
+            self.mask_radius = mask_radius # radius of the circular region used of the display of the SLM
+            
+            if self.mask_radius is not  None:
+                self.mask = display_reduction(dimension= self.mask_radius,value = 0)
+                self.mask = self.mask.astype('int8')
+                if wav_design_um is not None:
+                    riscaled_value = int((2**bitdepth)*(1 - (wav_um/wav_design_um)))
+                    if riscaled_value<0:
+                        riscaled_value=0
+                    self.mask_correction = display_reduction(dimension= self.mask_radius,value = riscaled_value)
+                    self.mask_correction = self.mask_correction.astype('int8')
+                else:
+                    self.mask_correction = np.zeros([self.height,self.width])
+                    self.mask_correction = self.mask_correction.astype('int8')

See comment above about this being superclass functionality.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.dx = pitch_um[0] / self.wav_um
+            self.dy = pitch_um[1] / self.wav_um

Removed in 0.1.0



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            # Make normalized coordinate grids.
+            xpix = (self.width - 1) *  np.linspace(-.5, .5, self.width)
+            ypix = (self.height - 1) * np.linspace(-.5, .5, self.height)
+            self.x_grid, self.y_grid = np.meshgrid(self.dx * xpix, self.dy * ypix)

Handled by superclass constructor.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            
+            
+          
+            super().__init__(
+                resolution = resolution,
+                bitdepth = bitdepth,
+                name = name ,
+                wav_um = wav_um,
+                wav_design_um = wav_design_um,
+                pitch_um = pitch_um,
+                settle_time_s = settle_time_s
+                )
+
+
+
+    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):

Let's remove the critical values. The user can check this outside this function if desired.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def memory_status(self):
+        self.occupied_slot_list = []
+        self.occupied_slot = 0
+        
+        for i in range(self.frame_mem_dim):
+            if self.memory[i]==1:
+                self.occupied_slot_list.append(i)
+                self.occupied_slot+=1
+        print('Slot occupied:', self.occupied_slot)
+        print('Occupied slots list:',self.occupied_slot_list)   

slmsuite does not currently support this sort of hardware frame caching, even though many SLMs support it. Let's keep this in for now, but know that the implementation will change when the superclass supports frame caching and we write compatible methods for other SLMs (Santec, Meadowlark, ...).



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def show_display(self):
+        dislpay_pattern = ham.Check_Disp_IMG(bID = self.ID, x_pixel=self.width, y_pixel = self.height,plot = True)
+        self.dislpay = np.array(dislpay_pattern)
+        

Remove. Covered by superclass .plot().



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def info(self,verbose = True):
+        print(self.name,'info:\n')
+        print('Head serial number:',self.head_serial_numb)
+        print('Head temperature:',self.head_temperature.value,' Celsius')
+        print('Controller temperature:',self.controller_temperature.value,' Celsius')
+        print('Led status:',self.led_status)
+        print('Slot occupied:', self.occupied_slot)
+        print('Occupied slots list:',self.occupied_slot_list)
+        print('Memory slot frame on the dislpay:',self.display_memory_slot)
+        if verbose == True:
+            self.check_dislpay()

Behavior on other SLMs is for info() to be a static method (callable without having to instantiate a class) to figure out what hardware is detected by an SDK. See how other SLM classes handle this.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def write(
+        self,
+        phase,
+        phase_correct=True,
+        blaze = True,
+        settle=True,
+        slot_number = 0
+    ):

I think the only change here is slot_number. Let's modify superclass write to accept **kwargs and then pass these **kwargs to the _write_hw() method. Also, note that write() was renamed set_phase (with backwards compatibility) in 0.2.0, I think. See how template.py suggests things.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +        ham.Check_FMem_Slot(bID = self.ID, x_pixel = self.width, y_pixel = self.height, slot_number = slot_number,plot = True)
+
+    
+
+    def write_bmp(self,file_path,slot_number):
+        '''
+        This method save a phase pattern in memory and then display it on the display of the SLM.
+        The input pattern must be in bmp format.
+        '''
+
+        ham.Write_FMemBMPPath(bID = self.ID,file_path = file_path,slot_number = slot_number,plot=False)
+        self.memory[slot_number] =1
+        
+     
+        
+    def load_vendor_phase_correction(self, file_path):

This should live in the superclass. See how template.py handles things.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def reboot(self):
+        ham.Reboot(self.ID)
+   
+    
+    
+    def close(self):
+        ham.Close_Device(bID_list = self.IDs, bID_size = self.n_dev)
+        
+      
+        
+      
+        
+
+# Function that reduce the SLM display active area to a circular region with the specified radius 
+
+def display_reduction(

Just to note, using for loops over i,j is very inefficient. Numpy allows you to use vectorized operations. For instance, see how analysis.fitfunctions.tophat2d() implements what you're doing here.



On slmsuite/hardware/slms/hamamatsu_slm_interpreter.py:
Let's move the methods in this file directly into the Hamamatsu class, whether as class or static methods.



In slmsuite/hardware/slms/hamamatsu_slm_interpreter.py:
> +# 1
+'''
+Establishes the communication with all the LCOS SLM controllers connected to the USB. 
+
+NB: make sure that the bID_list has the same lenght of the number of
+    connected devices to avoid problem with other functions.
+'''
+def Open_Device(bID_size = 1):
+    open_dev = Lcoslib.Open_Dev
+    open_dev.argtyes = [c_uint8*bID_size,c_int32]
+    open_dev.restype = c_int
+    array =c_uint8*bID_size
+    ID_list = array(0)
+    conn_dev = open_dev(byref(ID_list),bID_size)
+    if conn_dev == 0:
+        print("CONNECTION FAILED !!!")

Let's turn these print statements into errors: raise RuntimeError("Connection to Hamamatsu failed!")



On slmsuite/hardware/slms/L55-0491_USB_Control_SDK.pdf:
To reduce package file size, we have not been including SDK documentation in the package. Perhaps we can make a public Google Drive with various SDKs.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            
+            
+          
+            super().__init__(
+                resolution = resolution,
+                bitdepth = bitdepth,
+                name = name ,
+                wav_um = wav_um,
+                wav_design_um = wav_design_um,
+                pitch_um = pitch_um,
+                settle_time_s = settle_time_s
+                )
+
+
+
+    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):

This function should return the temps in a length-2 tuple.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ninovarde commented:
﻿Dear Ian,I’ve added the file to the previous pull request. Please let me know if I did it correctly, as I’m not very familiar with using GitHub.Best regards,Antonino VardèIl giorno mar 8 apr 2025 alle ore 19:12 ANTONINO VARDE' ***@***.***> ha scritto:Dear Ian,I’ve modified the script based on your comments, and now there is just one file called "hamamatsu.py.". Unfortunately, I am unable to check the script’s validity after the modifications since I’m no longer in the lab, due to the fact that I have completed my master's degree several weeks ago.Regarding the SDK, I think that it would be helpful to include the documentation on which the script is based. Do you already have a Google Drive folder for this purpose?Lastly, to share the updated version with you, do you think it would be better to create a new pull request and close the previous one, or is there a way to add the file to the existing pull request?Best regards,Antonino VardèIl giorno ven 4 apr 2025 alle ore 15:00 ANTONINO VARDE' ***@***.***> ha scritto:Dear Ian,Thank you for your comments. I will carefully read them and make the necessary adjustments to the scripts. I’ll update you in a couple of days, as I’m currently not able to spend much time at the computer.Thank you for your time.Best regards,Antonino VardèSent from my iPhoneOn 3 Apr 2025, at 21:05, Ian Christen ***@***.***> wrote:﻿
@ichristen commented on this pull request.

Hi @ninovarde , please see my comments.
As a general comment, any public functions will also need to have docstrings in the style of the rest of the package. In part because of this, my preference is to remove repeated or superfluous functions and focus on comprehensively documenting the functions needed to implement the SDK.
Let me know what you think!

In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            blaze_angle = None, # in degree
+            mask_radius = None # in pixel

If we add these arguments, it should be in the superclass SLM. This is already sortof implementable by setting the phase correction to a blaze. Would have to check, but setting part of the phase correction to np.nan might play the same role as the mask.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            frame_mem_dim = 819,
+            head_serial_numb = 'LSH0804453\x00',
+            bID_size = 1,

Let's try to keep the number of arguments down to the minimum. Otherwise, the package becomes unmaintainable. See other SLM classes for what minimum looks like.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.width = resolution[0]
+            self.height = resolution[1]

Removed in 0.1.0



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            temp =  ham.Check_Temp(self.ID)
+            self.head_temperature = temp[0]
+            self.controller_temperature = temp[1]
+            
+            self.led_status = ham.Check_LED(self.ID)

Let's not have static cached attributes here. Instead, the user should poll the hardware for fresh variables.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.blaze_angle = blaze_angle # angle at which the produced patterns are shifted from the zero order
+        
+            self.mask_radius = mask_radius # radius of the circular region used of the display of the SLM
+            
+            if self.mask_radius is not  None:
+                self.mask = display_reduction(dimension= self.mask_radius,value = 0)
+                self.mask = self.mask.astype('int8')
+                if wav_design_um is not None:
+                    riscaled_value = int((2**bitdepth)*(1 - (wav_um/wav_design_um)))
+                    if riscaled_value<0:
+                        riscaled_value=0
+                    self.mask_correction = display_reduction(dimension= self.mask_radius,value = riscaled_value)
+                    self.mask_correction = self.mask_correction.astype('int8')
+                else:
+                    self.mask_correction = np.zeros([self.height,self.width])
+                    self.mask_correction = self.mask_correction.astype('int8')

See comment above about this being superclass functionality.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            self.dx = pitch_um[0] / self.wav_um
+            self.dy = pitch_um[1] / self.wav_um

Removed in 0.1.0



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            # Make normalized coordinate grids.
+            xpix = (self.width - 1) *  np.linspace(-.5, .5, self.width)
+            ypix = (self.height - 1) * np.linspace(-.5, .5, self.height)
+            self.x_grid, self.y_grid = np.meshgrid(self.dx * xpix, self.dy * ypix)

Handled by superclass constructor.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            
+            
+          
+            super().__init__(
+                resolution = resolution,
+                bitdepth = bitdepth,
+                name = name ,
+                wav_um = wav_um,
+                wav_design_um = wav_design_um,
+                pitch_um = pitch_um,
+                settle_time_s = settle_time_s
+                )
+
+
+
+    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):

Let's remove the critical values. The user can check this outside this function if desired.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def memory_status(self):
+        self.occupied_slot_list = []
+        self.occupied_slot = 0
+        
+        for i in range(self.frame_mem_dim):
+            if self.memory[i]==1:
+                self.occupied_slot_list.append(i)
+                self.occupied_slot+=1
+        print('Slot occupied:', self.occupied_slot)
+        print('Occupied slots list:',self.occupied_slot_list)   

slmsuite does not currently support this sort of hardware frame caching, even though many SLMs support it. Let's keep this in for now, but know that the implementation will change when the superclass supports frame caching and we write compatible methods for other SLMs (Santec, Meadowlark, ...).



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def show_display(self):
+        dislpay_pattern = ham.Check_Disp_IMG(bID = self.ID, x_pixel=self.width, y_pixel = self.height,plot = True)
+        self.dislpay = np.array(dislpay_pattern)
+        

Remove. Covered by superclass .plot().



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def info(self,verbose = True):
+        print(self.name,'info:\n')
+        print('Head serial number:',self.head_serial_numb)
+        print('Head temperature:',self.head_temperature.value,' Celsius')
+        print('Controller temperature:',self.controller_temperature.value,' Celsius')
+        print('Led status:',self.led_status)
+        print('Slot occupied:', self.occupied_slot)
+        print('Occupied slots list:',self.occupied_slot_list)
+        print('Memory slot frame on the dislpay:',self.display_memory_slot)
+        if verbose == True:
+            self.check_dislpay()

Behavior on other SLMs is for info() to be a static method (callable without having to instantiate a class) to figure out what hardware is detected by an SDK. See how other SLM classes handle this.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def write(
+        self,
+        phase,
+        phase_correct=True,
+        blaze = True,
+        settle=True,
+        slot_number = 0
+    ):

I think the only change here is slot_number. Let's modify superclass write to accept **kwargs and then pass these **kwargs to the _write_hw() method. Also, note that write() was renamed set_phase (with backwards compatibility) in 0.2.0, I think. See how template.py suggests things.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +        ham.Check_FMem_Slot(bID = self.ID, x_pixel = self.width, y_pixel = self.height, slot_number = slot_number,plot = True)
+
+    
+
+    def write_bmp(self,file_path,slot_number):
+        '''
+        This method save a phase pattern in memory and then display it on the display of the SLM.
+        The input pattern must be in bmp format.
+        '''
+
+        ham.Write_FMemBMPPath(bID = self.ID,file_path = file_path,slot_number = slot_number,plot=False)
+        self.memory[slot_number] =1
+        
+     
+        
+    def load_vendor_phase_correction(self, file_path):

This should live in the superclass. See how template.py handles things.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +    def reboot(self):
+        ham.Reboot(self.ID)
+   
+    
+    
+    def close(self):
+        ham.Close_Device(bID_list = self.IDs, bID_size = self.n_dev)
+        
+      
+        
+      
+        
+
+# Function that reduce the SLM display active area to a circular region with the specified radius 
+
+def display_reduction(

Just to note, using for loops over i,j is very inefficient. Numpy allows you to use vectorized operations. For instance, see how analysis.fitfunctions.tophat2d() implements what you're doing here.



On slmsuite/hardware/slms/hamamatsu_slm_interpreter.py:
Let's move the methods in this file directly into the Hamamatsu class, whether as class or static methods.



In slmsuite/hardware/slms/hamamatsu_slm_interpreter.py:
> +# 1
+'''
+Establishes the communication with all the LCOS SLM controllers connected to the USB. 
+
+NB: make sure that the bID_list has the same lenght of the number of
+    connected devices to avoid problem with other functions.
+'''
+def Open_Device(bID_size = 1):
+    open_dev = Lcoslib.Open_Dev
+    open_dev.argtyes = [c_uint8*bID_size,c_int32]
+    open_dev.restype = c_int
+    array =c_uint8*bID_size
+    ID_list = array(0)
+    conn_dev = open_dev(byref(ID_list),bID_size)
+    if conn_dev == 0:
+        print("CONNECTION FAILED !!!")

Let's turn these print statements into errors: raise RuntimeError("Connection to Hamamatsu failed!")



On slmsuite/hardware/slms/L55-0491_USB_Control_SDK.pdf:
To reduce package file size, we have not been including SDK documentation in the package. Perhaps we can make a public Google Drive with various SDKs.



In slmsuite/hardware/slms/hamamatsu_slm_class.py:
> +            
+            
+          
+            super().__init__(
+                resolution = resolution,
+                bitdepth = bitdepth,
+                name = name ,
+                wav_um = wav_um,
+                wav_design_um = wav_design_um,
+                pitch_um = pitch_um,
+                settle_time_s = settle_time_s
+                )
+
+
+
+    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):

This function should return the temps in a length-2 tuple.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ninovarde commented:
Dear Ian,

Did you have a chance to look at the script? If so, do you have any
comments on it?

Best wishes,
Antonino Vardè

Il giorno mar 8 apr 2025 alle ore 19:55 ANTONINO VARDE' <
***@***.***> ha scritto:

> ﻿
> Dear Ian,
>
> I’ve added the file to the previous pull request. Please let me know if I
> did it correctly, as I’m not very familiar with using GitHub.
>
> Best regards,
> Antonino Vardè
>
> Il giorno mar 8 apr 2025 alle ore 19:12 ANTONINO VARDE' <
> ***@***.***> ha scritto:
>
>> Dear Ian,
>>
>> I’ve modified the script based on your comments, and now there is just
>> one file called "hamamatsu.py.". Unfortunately, I am unable to check the
>> script’s validity after the modifications since I’m no longer in the lab,
>> due to the fact that I have completed my master's degree several weeks ago.
>>
>> Regarding the SDK, I think that it would be helpful to include the
>> documentation on which the script is based. Do you already have a Google
>> Drive folder for this purpose?
>>
>> Lastly, to share the updated version with you, do you think it would be
>> better to create a new pull request and close the previous one, or is there
>> a way to add the file to the existing pull request?
>>
>> Best regards,
>> Antonino Vardè
>>
>> Il giorno ven 4 apr 2025 alle ore 15:00 ANTONINO VARDE' <
>> ***@***.***> ha scritto:
>>
>>> Dear Ian,
>>> Thank you for your comments. I will carefully read them and make the
>>> necessary adjustments to the scripts. I’ll update you in a couple of days,
>>> as I’m currently not able to spend much time at the computer.
>>> Thank you for your time.
>>>
>>> Best regards,
>>> Antonino Vardè
>>> Sent from my iPhone
>>>
>>> On 3 Apr 2025, at 21:05, Ian Christen ***@***.***> wrote:
>>>
>>> ﻿
>>>
>>> ***@***.**** commented on this pull request.
>>>
>>> Hi @ninovarde <https://github.com/ninovarde> , please see my comments.
>>>
>>> As a general comment, any public functions will also need to have
>>> docstrings in the style of the rest of the package. In part because of
>>> this, my preference is to remove repeated or superfluous functions and
>>> focus on comprehensively documenting the functions needed to implement the
>>> SDK.
>>>
>>> Let me know what you think!
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027558388>:
>>>
>>> > +            blaze_angle = None, # in degree
>>> +            mask_radius = None # in pixel
>>>
>>> If we add these arguments, it should be in the superclass SLM. This is
>>> already sortof implementable by setting the phase correction to a blaze.
>>> Would have to check, but setting part of the phase correction to np.nan
>>> might play the same role as the mask.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027559745>:
>>>
>>> > +            frame_mem_dim = 819,
>>> +            head_serial_numb = 'LSH0804453\x00',
>>> +            bID_size = 1,
>>>
>>> Let's try to keep the number of arguments down to the minimum.
>>> Otherwise, the package becomes unmaintainable. See other SLM classes for
>>> what minimum looks like.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027560437>:
>>>
>>> > +            self.width = resolution[0]
>>> +            self.height = resolution[1]
>>>
>>> Removed in 0.1.0
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027562091>:
>>>
>>> > +            temp =  ham.Check_Temp(self.ID)
>>> +            self.head_temperature = temp[0]
>>> +            self.controller_temperature = temp[1]
>>> +
>>> +            self.led_status = ham.Check_LED(self.ID)
>>>
>>> Let's not have static cached attributes here. Instead, the user should
>>> poll the hardware for fresh variables.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027562865>:
>>>
>>> > +            self.blaze_angle = blaze_angle # angle at which the produced patterns are shifted from the zero order
>>> +
>>> +            self.mask_radius = mask_radius # radius of the circular region used of the display of the SLM
>>> +
>>> +            if self.mask_radius is not  None:
>>> +                self.mask = display_reduction(dimension= self.mask_radius,value = 0)
>>> +                self.mask = self.mask.astype('int8')
>>> +                if wav_design_um is not None:
>>> +                    riscaled_value = int((2**bitdepth)*(1 - (wav_um/wav_design_um)))
>>> +                    if riscaled_value<0:
>>> +                        riscaled_value=0
>>> +                    self.mask_correction = display_reduction(dimension= self.mask_radius,value = riscaled_value)
>>> +                    self.mask_correction = self.mask_correction.astype('int8')
>>> +                else:
>>> +                    self.mask_correction = np.zeros([self.height,self.width])
>>> +                    self.mask_correction = self.mask_correction.astype('int8')
>>>
>>> See comment above about this being superclass functionality.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027563270>:
>>>
>>> > +            self.dx = pitch_um[0] / self.wav_um
>>> +            self.dy = pitch_um[1] / self.wav_um
>>>
>>> Removed in 0.1.0
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027563651>:
>>>
>>> > +            # Make normalized coordinate grids.
>>> +            xpix = (self.width - 1) *  np.linspace(-.5, .5, self.width)
>>> +            ypix = (self.height - 1) * np.linspace(-.5, .5, self.height)
>>> +            self.x_grid, self.y_grid = np.meshgrid(self.dx * xpix, self.dy * ypix)
>>>
>>> Handled by superclass constructor.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027564194>:
>>>
>>> > +
>>> +
>>> +
>>> +            super().__init__(
>>> +                resolution = resolution,
>>> +                bitdepth = bitdepth,
>>> +                name = name ,
>>> +                wav_um = wav_um,
>>> +                wav_design_um = wav_design_um,
>>> +                pitch_um = pitch_um,
>>> +                settle_time_s = settle_time_s
>>> +                )
>>> +
>>> +
>>> +
>>> +    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):
>>>
>>> Let's remove the critical values. The user can check this outside this
>>> function if desired.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027567136>:
>>>
>>> > +    def memory_status(self):
>>> +        self.occupied_slot_list = []
>>> +        self.occupied_slot = 0
>>> +
>>> +        for i in range(self.frame_mem_dim):
>>> +            if self.memory[i]==1:
>>> +                self.occupied_slot_list.append(i)
>>> +                self.occupied_slot+=1
>>> +        print('Slot occupied:', self.occupied_slot)
>>> +        print('Occupied slots list:',self.occupied_slot_list)
>>>
>>> slmsuite does not currently support this sort of hardware frame caching,
>>> even though many SLMs support it. Let's keep this in for now, but know that
>>> the implementation will change when the superclass supports frame caching
>>> and we write compatible methods for other SLMs (Santec, Meadowlark, ...).
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027567582>:
>>>
>>> > +    def show_display(self):
>>> +        dislpay_pattern = ham.Check_Disp_IMG(bID = self.ID, x_pixel=self.width, y_pixel = self.height,plot = True)
>>> +        self.dislpay = np.array(dislpay_pattern)
>>> +
>>>
>>> Remove. Covered by superclass .plot().
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027568694>:
>>>
>>> > +    def info(self,verbose = True):
>>> +        print(self.name,'info:\n')
>>> +        print('Head serial number:',self.head_serial_numb)
>>> +        print('Head temperature:',self.head_temperature.value,' Celsius')
>>> +        print('Controller temperature:',self.controller_temperature.value,' Celsius')
>>> +        print('Led status:',self.led_status)
>>> +        print('Slot occupied:', self.occupied_slot)
>>> +        print('Occupied slots list:',self.occupied_slot_list)
>>> +        print('Memory slot frame on the dislpay:',self.display_memory_slot)
>>> +        if verbose == True:
>>> +            self.check_dislpay()
>>>
>>> Behavior on other SLMs is for info() to be a static method (callable
>>> without having to instantiate a class) to figure out what hardware is
>>> detected by an SDK. See how other SLM classes handle this.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027571087>:
>>>
>>> > +    def write(
>>> +        self,
>>> +        phase,
>>> +        phase_correct=True,
>>> +        blaze = True,
>>> +        settle=True,
>>> +        slot_number = 0
>>> +    ):
>>>
>>> I think the only change here is slot_number. Let's modify superclass
>>> write to accept **kwargs and then pass these **kwargs to the _write_hw()
>>> method. Also, note that write() was renamed set_phase (with backwards
>>> compatibility) in 0.2.0, I think. See how template.py suggests things.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027572158>:
>>>
>>> > +        ham.Check_FMem_Slot(bID = self.ID, x_pixel = self.width, y_pixel = self.height, slot_number = slot_number,plot = True)
>>> +
>>> +
>>> +
>>> +    def write_bmp(self,file_path,slot_number):
>>> +        '''
>>> +        This method save a phase pattern in memory and then display it on the display of the SLM.
>>> +        The input pattern must be in bmp format.
>>> +        '''
>>> +
>>> +        ham.Write_FMemBMPPath(bID = self.ID,file_path = file_path,slot_number = slot_number,plot=False)
>>> +        self.memory[slot_number] =1
>>> +
>>> +
>>> +
>>> +    def load_vendor_phase_correction(self, file_path):
>>>
>>> This should live in the superclass. See how template.py handles things.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027576571>:
>>>
>>> > +    def reboot(self):
>>> +        ham.Reboot(self.ID)
>>> +
>>> +
>>> +
>>> +    def close(self):
>>> +        ham.Close_Device(bID_list = self.IDs, bID_size = self.n_dev)
>>> +
>>> +
>>> +
>>> +
>>> +
>>> +
>>> +# Function that reduce the SLM display active area to a circular region with the specified radius
>>> +
>>> +def display_reduction(
>>>
>>> Just to note, using for loops over i,j is very inefficient. Numpy allows
>>> you to use vectorized operations. For instance, see how
>>> analysis.fitfunctions.tophat2d() implements what you're doing here.
>>> ------------------------------
>>>
>>> On slmsuite/hardware/slms/hamamatsu_slm_interpreter.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027577304>:
>>>
>>> Let's move the methods in this file directly into the Hamamatsu class,
>>> whether as class or static methods.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_interpreter.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027578621>:
>>>
>>> > +# 1
>>> +'''
>>> +Establishes the communication with all the LCOS SLM controllers connected to the USB.
>>> +
>>> +NB: make sure that the bID_list has the same lenght of the number of
>>> +    connected devices to avoid problem with other functions.
>>> +'''
>>> +def Open_Device(bID_size = 1):
>>> +    open_dev = Lcoslib.Open_Dev
>>> +    open_dev.argtyes = [c_uint8*bID_size,c_int32]
>>> +    open_dev.restype = c_int
>>> +    array =c_uint8*bID_size
>>> +    ID_list = array(0)
>>> +    conn_dev = open_dev(byref(ID_list),bID_size)
>>> +    if conn_dev == 0:
>>> +        print("CONNECTION FAILED !!!")
>>>
>>> Let's turn these print statements into errors: raise
>>> RuntimeError("Connection to Hamamatsu failed!")
>>> ------------------------------
>>>
>>> On slmsuite/hardware/slms/L55-0491_USB_Control_SDK.pdf
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027579914>:
>>>
>>> To reduce package file size, we have not been including SDK
>>> documentation in the package. Perhaps we can make a public Google Drive
>>> with various SDKs.
>>> ------------------------------
>>>
>>> In slmsuite/hardware/slms/hamamatsu_slm_class.py
>>> <https://github.com/slmsuite/slmsuite/pull/100#discussion_r2027564695>:
>>>
>>> > +
>>> +
>>> +
>>> +            super().__init__(
>>> +                resolution = resolution,
>>> +                bitdepth = bitdepth,
>>> +                name = name ,
>>> +                wav_um = wav_um,
>>> +                wav_design_um = wav_design_um,
>>> +                pitch_um = pitch_um,
>>> +                settle_time_s = settle_time_s
>>> +                )
>>> +
>>> +
>>> +
>>> +    def check_temperature(self,critical_head_value = 100,critical_controller_value = 100):
>>>
>>> This function should return the temps in a length-2 tuple.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/slmsuite/slmsuite/pull/100#pullrequestreview-2740830245>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/BGWNSXQMY6AOPKGU2L2MUVT2XWA7DAVCNFSM6AAAAAB2M54DSSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONBQHAZTAMRUGU>
>>> .
>>> You are receiving this because you were mentioned.Message ID:
>>> ***@***.***>
>>>
>>>

ichristen commented:
Sorry for my late reply. I'll take a look this weekend. Thank you for your work on this!

ninovarde commented:
Don’t worry. Let me know what you think.Best wishes,AntoninoOn 24 Apr 2025, at 20:50, Ian Christen ***@***.***> wrote:﻿
ichristen left a comment (slmsuite/slmsuite#100)
Sorry for my late reply. I'll take a look this weekend. Thank you for your work on this!

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ichristen commented:
Hi @ninovarde 

Sorry for my delay on this. I made some edits to your pull request. I don't have push permissions to your fork ninovarde/slmsuite, so I moved my changes to the local [slmsuite/nino_varde_slm_hamamatsu](https://github.com/slmsuite/slmsuite/tree/nino_varde_slm_hamamatsu) branch.

Changes:

- I removed files other than `hamamatsu.py`.
- I removed superclass functions from `Hamamatsu`, but I added a new feature so you can still use `.set_phase(phase, slot_number=0)` by passing `**kwargs`.
- I cleaned up some of the functions/style to match with slmsuite.
- I removed many of the constructor arguments (which are passed as `**kwargs`).
- I removed hardcoding for the .dll path (this is untested!).
- I added functionality to check and set the SLM mode (DVI vs USB). My understanding is that the class you wrote is designed to work with USB. For DVI, I think `ScreenMirrored` is appropriate?

You mentioned before that you don't have access to the SLM to test. If this is still the case, I'm alright with pushing the code with the **(Untested)** warning, alongside mentioning that DVI mode with `ScreenMirrored` is an option. Does this sound reasonable?

Best,
Ian

ninovarde commented:
Dear Ian,

The code seems great! I still have a lot to learn, and for this reason, I
hope to be able to collaborate more with you in the future.

As you said, the code was designed for USB mode.

Regarding the possibility of testing it, if someone uses it in the future
and finds some problems, you can put them in contact with me, and I will be
happy to help them (Although I believe GitHub should notify me if someone
opens a related issue, right?). I don't know when they plan to implement
the SLM in the main setup in my old laboratory, but once they do and test
the library, I’ll let you know.


It's been a pleasure,

Best regards,
Antonino Vardè




Il giorno sab 3 mag 2025 alle ore 03:41 Ian Christen <
***@***.***> ha scritto:

> *ichristen* left a comment (slmsuite/slmsuite#100)
> <https://github.com/slmsuite/slmsuite/pull/100#issuecomment-2848353008>
>
> Hi @ninovarde <https://github.com/ninovarde>
>
> Sorry for my delay on this. I made some edits to your pull request. I
> don't have push permissions to your fork ninovarde/slmsuite, so I moved my
> changes to the local slmsuite/nino_varde_slm_hamamatsu
> <https://github.com/slmsuite/slmsuite/tree/nino_varde_slm_hamamatsu>
> branch.
>
> Changes:
>
>    - I removed files other than hamamatsu.py.
>    - I removed superclass functions from Hamamatsu, but I added a new
>    feature so you can still use .set_phase(phase, slot_number=0) by
>    passing **kwargs.
>    - I cleaned up some of the functions/style to match with slmsuite.
>    - I removed many of the constructor arguments (which are passed as
>    **kwargs).
>    - I removed hardcoding for the .dll path (this is untested!).
>    - I added functionality to check and set the SLM mode (DVI vs USB). My
>    understanding is that the class you wrote is designed to work with USB. For
>    DVI, I think ScreenMirrored is appropriate?
>
> You mentioned before that you don't have access to the SLM to test. If
> this is still the case, I'm alright with pushing the code with the
> *(Untested)* warning, alongside mentioning that DVI mode with
> ScreenMirrored is an option. Does this sound reasonable?
>
> Best,
> Ian
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/slmsuite/slmsuite/pull/100#issuecomment-2848353008>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BGWNSXVLEDVTF54O4ZNZ3VT24QNDZAVCNFSM6AAAAAB2M54DSSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNBYGM2TGMBQHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ichristen commented:
Sounds good! I'm closing this pull request in favor of #106
